qualtran/surface_code/ui.py

```diff
@@ -0,0 +1,608 @@
+        ]
+        memory_footprint['qubits'] = [res.footprint - factory.footprint(), factory.footprint()]
+        fig = px.pie(
+            memory_footprint, values='qubits', names='source', title='Physical Qubit Utilization'
+        )
+        fig.update_traces(textinfo='value')
+        return fig
+    else:
+        factory = MultiFactory(magic_factory, int(magic_count))
+        memory_footprint = pd.DataFrame(columns=['source', 'qubits'])
+        memory_footprint['source'] = [
+            'logical qubits + routing overhead',
+            'Magic State Distillation',
+        ]
+        memory_footprint['qubits'] = [logical_qubits(algorithm), factory.footprint()]
+        fig = px.pie(
+            memory_footprint, values='qubits', names='source', title='Physical Qubit Utilization'
+        )
+        fig.update_traces(textinfo='value')
+        return fig
+
+
+def format_duration(duration: Sequence[float]) -> Tuple[str, Sequence[float]]:
+    """Returns a tuple of the format (unit, duration)
+
+    Finds the best unit to report `duration` and assumes that `duration` is initially in us.
+    """
+    unit = 'us'
+    if duration[0] > 1000:
+        duration = [d / 1000 for d in duration]
+        unit = 'ms'
+    if duration[0] > 1000:
+        duration = [d / 1000 for d in duration]
+        unit = 's'
+    if duration[0] > 60:
+        duration = [d / 60 for d in duration]
+        unit = 'min'
+    if duration[0] > 60:
+        duration = [d / 60 for d in duration]
+        unit = 'hours'
+    if duration[0] > 24:
+        duration = [d / 24 for d in duration]
+        unit = 'days'
+    return unit, duration
+
+
+def create_runtime_plot(
+    physical_error_rate: float,
+    error_budget: float,
+    estimation_model: str,
+    algorithm: AlgorithmSummary,
+    qec: qecs.QuantumErrorCorrectionSchemeSummary,
+    magic_factory: magic_state_factory.MagicStateFactory,
+    magic_count: int,
+    rotation_model: rotation_cost_model.RotationCostModel,
+    needed_magic: MagicCount,
+) -> Tuple[Dict[str, Any], go.Figure]:
+    """Creates the runtime figure and decides whether to display it or not.
+
+    Currently displays the runtime plot for the Beverland model only.
+    """
+    if estimation_model == 'GidneyFolwer (arxiv:1812.01238)':
+        return {'display': 'none'}, go.Figure()
+    factory = MultiFactory(magic_factory, int(magic_count))
+    c_min = minimum_time_steps(
+        error_budget=error_budget, alg=algorithm, rotation_model=rotation_model
+    )
+    factory_cycles = factory.n_cycles(needed_magic, physical_error_rate)
+    min_num_factories = int(np.ceil(factory_cycles / c_min))
+    magic_counts = list(
+        1 + np.random.choice(min_num_factories, replace=False, size=min(min_num_factories, 5))
+    )
+    magic_counts.sort(reverse=True)
+    magic_counts = np.array(magic_counts)
+    time_steps = np.ceil(factory_cycles / magic_counts)
+    magic_counts[0] = min_num_factories
+    time_steps[0] = c_min
```

Comment:
Ideally, we should define a dataclass that holds all the information required for the pie chart and runtime plots of the UI and then have methods defined in the corresponding estimation model files that return this dataclass for a given list of input parameters. We shouldn't have any logic to compute values for estimation models (eg: `total_magic` for gidney model; which simply returns `n_t / 4 + n_ccz`. This is an approximation because it's trivial to decompose 1 CCZ into 4 Ts but combining 4Ts in a logical circuit is non-trivial and cannot always be done) in the UI file. 

The UI file should strictly delegate all computation and preferably call just one function to get a populated data class which can then be plotted.

Reply:
I agree, but this is a first step. I plan to add an abstraction to resource estimation models that covers both of the models we have and make it easier to add others in the future. but this can be in followup PRs